### ISSUE TITLE
Performance optimisation 

### ISSUE THREAD

**Issue by bodinsamuel** [algolia/npm-search#372]
## `db.allDocs`

This is probably the slowest operation for the moment and the good news it is possible to improve. The payload is big and the API respond almost instantly, it's mostly Idle CPU and a big parsing time at the end. 

If we wrap the call into an in memory store (either in the same thread or in a fork) we could always fetch in the background a few thousand rows that way the next `loop()` does not wait at all.

**Cons:**
- Forking will mean serialize/unserialize between thread, not sure we will gain that much perf.
- In memory will mean that we can not prefetch too much.

(There is one problem that can be solved in some other way is that we are downloading all versions of every packages, that means all html readme too. This can be pretty big for some package. We could find a way to get `latest` only and fetch the list of versions with an other endpoint/way)

## `getDownloads`

Not as slow as imagined but takes 20-30% of `saveDocs` duration. It's not a critical info (even though it's part of the UX of the website), it could be done in a separate process.

## `getTSSupport`

Might be the slowest metadata operation. Between 3s - 7s for 100 packages. Account for > 50% of the `saveDocs` duration.
Sometimes it can can stuck for super long time, blocking everything.
<img width="602" alt="Screenshot 2019-08-03 18 56 04" src="https://user-images.githubusercontent.com/1637651/62414761-5e5ef180-b620-11e9-887d-99e967c9312b.png">

## `getChangelogs`

Is surprisingly not as slow as expected. Probably because we are matching CHANGELOG.md almost everytime. This could be done in an other process too.
Account for 20-30% of the `saveDocs` duration.

**Ref**
(except formatPackages, all timing on the left of a charts are in seconds)
<img width="1217" alt="Screenshot 2019-08-03 18 48 50" src="https://user-images.githubusercontent.com/1637651/62414736-eee90200-b61f-11e9-9574-2fb97beec2ed.png">

**Comment by bodinsamuel** [comment]
`db.allDocs` followup.

Tried to implement a basic prefetch strategy, in the same process, it seems to improve perf x2. I can't even prefetch fast enough. This is good news 💪

**Comment by Haroenv** [comment]
I think typescript jumps so much because it depends which kind of package it is how many requests it does. If included it doesn’t do requests, otherwise it checks @types via npm, and finally it uses unpkg if the default files are maybe d.ts. As said, we can see if jsdelivr or unpkg can give us all the files and if that’s faster

**Comment by bodinsamuel** [comment]
Update 2021:

Not much has changed `db.list` is still slow (100ms to 2s) but the main culprit is having every versions inside the document.
It tops up the ram pretty quick, we currently can only bootstrap 10 by 10 max.

I have tried to do `db.list({ include_docs: false })` but requires 2 additional calls to complete the document.
- First for the `latest` version that contains most the info
- Second for the list of versions

With this we are still missing the prop `time`; we do have the list of versions but not the created time. 
If I had to decide I would definitely go the for First call only and be done with it, but yarnpkg does display the list of versions with date of creation so that would be breaking change.

--- 

My only lead at this time is to move from Heroku to Compute Engine with a e2-standard-2 preemptible, that would give us more 2x CPU and 8x RAM for half the price. 
In the end it does not solve the issue that some packages are too big with all the versions, but it will give us some time.
